### PR TITLE
Explicitly declare ctypes variables in Session methods for every parameter to be passed into Library

### LINKED ISSUE #443: Explicitly declare all ctypes functions passed to Library
In Session.py, we call Library using Python types such as int or float, and then let ctypes magically convert to the proper type based on information configured in Library.py (argtypes).

This works, but obscures errors and also makes unit testing less precise because expectations aren't set in terms of the actual things that are passed at runtime, and by mocking Library we skip all the implicit conversions happening in the real Library.

Change Session.py so that we explicitly declare all ctype objects that are passed to Library.

### LINKED ISSUE #505: Input lists don't work
Some functions (notably in the NI-FGEN API) require client to pass in an array / list. Our generator doesn't handle it correctly. We just turn around and hand it off to the Library class. 

See this error I got while writing an NI-DCPower example (attribute_ids is a Python list):

```
Z:\>python src\nidcpower\examples\nidcpower_advanced_sequence.py
Traceback (most recent call last):
  File "src\nidcpower\examples\nidcpower_advanced_sequence.py", line 31, in <module>
    session.create_advanced_sequence('my_sequence', attribute_ids)
  File "C:\Python27\lib\site-packages\nidcpower\session.py", line 4338, in create_advanced_sequence
    error_code = self._library.niDCPower_CreateAdvancedSequence(self._vi, sequence_name.encode(self._encoding), attribute_id_count, attribute_ids, set_as_active_sequence)
  File "C:\Python27\lib\site-packages\nidcpower\library.py", line 142, in niDCPower_CreateAdvancedSequence
    return self.niDCPower_CreateAdvancedSequence_cfunc(vi, sequence_name, attribute_id_count, attribute_ids, set_as_active_sequence)
ctypes.ArgumentError: argument 4: <type 'exceptions.TypeError'>: expected LP_c_long instance instead of list
```

### LINKED ISSUE #523: NI-FAKE GetAStringWithSpecifiedMaximumSize function is invalid
We have some unit tests in NI-FAKE for a function GetAStringWithSpecifiedMaximumSize:

    niFake_GetAStringWithSpecifiedMaximumSize(ViSession, ViChar*, ViInt32);

The corresponding Session method:

    def get_a_string_with_specified_maximum_size(self, buffer_size):

The C user is supposed to pass a buffer and the max size of the buffer. In Python, the buffer_size is exposed. This makes no sense from the Python API point of view. Clients should never have to deal with string memory allocation.

Recommendation is to remove the function. There are other unit tests in which the function is used for different reasons - a different function should be used in those.

### PR DESCRIPTION
**Pull request by marcoskirsch**
[X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).

~~[ ] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~

### What does this Pull Request accomplish?

Adds code to Session so that all ctypes objects to be passed into Library function calls are instantiated in the scope of the method. Before this, we were only doing so for output parameters and buffers and letting ctypes handle the rest.

This forces us to write more correct unit tests because expectations on Library calls will need to be set using the proper types.

It also make type errors be caught earlier in Session, rather than wait for the inner magic of ctypes to raise an error; for example: when client passes a string where a float is expected.

It also fixes several cases we hadn't properly added support for.

### List issues fixed by this Pull Request below, if any.

* Fix #443 
* Fix #505 
* Fix #523 

### What testing has been done?

* tox
* nimi-bot (all system tests pass)

### CODE REVIEW

**Comment by marcoskirsch** [comment]
At this stage:

* All code appears to be correct
* All system tests pass, including those that have input buffers
* Lots of unit tests are commented out because they still need retrofits. Thus drastic decrease in code coverage.
* One unit test was retrofitted. I introduced a bunch of Matcher classes for setting strict expectations of ctype objects. More will be needed.

You can start reviewing the changes to generator and production code. I will keep working on the unit tests. You can look at that if you're curious or in order to provide early feedback.

File: src/nifake/tests/test_session.py [hunk 1]
```diff
@@ -1,12 +1,72 @@
...
+
+    def __eq__(self, other):
+        if not isinstance(other, self.expected_type):
+            print("Unexpected type. Expected: {0}. Received: {1}".format(self.expected_type, type(other)))
+            return False
+        if other.value == self.expected_value:
+            return True
+        else:
+            print("Unexpected value. Expected: {0}. Received: {1}".format(self.expected_value, other.value))
+        return
```
Comment (texasaggie97) [inline src/nifake/tests/test_session.py]: The other two code paths return a value. You should here as well.

File: src/nifake/tests/test_session.py [hunk 2]
```diff
@@ -1,12 +1,72 @@
...
+            print("Unexpected type. Expected: {0}. Received: {1}".format(ctypes.Array, type(other)))
+            return False
+        if len(other) < len(self.expected_string_value) + 1:  # +1 for NULL terminating character
+            print("Unexpected length in C string. Expected at least: {0}. Received {1}".format(len(other), len(self.expected_string_value) + 1))
+            return False
+        if other.value.decode("ascii") == self.expected_string_value:
+            return True
+        else:
+            print("Unexpected value. Expected {0}. Received: {1}".format(self.expected_string_value, other.value.decode))
+        return
```
Comment (texasaggie97) [inline src/nifake/tests/test_session.py]: Same thing about other code paths returning a value

File: src/nifake/tests/test_session.py [hunk 3]
```diff
@@ -1,12 +1,72 @@
...
+        else:
+            print("Unexpected value. Expected: {0}. Received: {1}".format(self.expected_value, other.value))
+        return
+
+
+class StringMatcher(object):
+    def __init__(self, expected_string_value):
+        self.expected_string_value = expected_string_value
+
+    def __eq__(self, other):
```
Comment (texasaggie97) [inline src/nifake/tests/test_session.py]: We know it is a string, should you also check that the element type is ViChar?

File: src/nifake/tests/test_session.py [hunk 4]
```diff
@@ -1,12 +1,72 @@
...
+# import sys
+from nifake import visatype
+# import warnings
 
-from mock import ANY
+# from mock import ANY
 from mock import patch
 
+
+# Matching classes for setting mock expectations:
```
Comment (texasaggie97) [inline src/nifake/tests/test_session.py]: Can/Should these matcher helpers be moved to somewhere else to single source them? Add them to mock_helper and rename it to unit_test_helper?

File: build/helper/codegen_helper.py [hunk 5]
```diff
@@ -111,24 +111,79 @@ def get_enum_type_check_snippet(parameter, indent):
     return enum_check
 
 
+def _get_buffer_parameter_for_size_parameter(parameter, parameters):
+    '''If parameter represents the size of another parameter in the C API, returns that other parameter. Otherwise None.'''
+    for p in parameters:
+        if p['is_buffer'] and p['size']['value'] == parameter['name']:
+            assert p['is_buffer'] is True, '{0} expected to be a buffer because its size is specified by {1}'.format(p['name'], parameter['name'])
```
[the remark this answers is not in the document]
Comment (marcoskirsch) [reply build/helper/codegen_helper.py]: duh!

File: build/helper/codegen_helper.py [hunk 6]
```diff
@@ -111,24 +111,79 @@ def get_enum_type_check_snippet(parameter, indent):
...
+    # Note that we append "# case x". It's ugly in the generated code but it's sooo useful for debugging code generation problems.
+    if parameter['direction'] == 'in':
+        if parameter['is_session_handle'] is True:
+            definition = 'visatype.{0}(self._{1})  # case 1'.format(parameter['ctypes_type'], parameter['python_name'])
+        elif parameter['is_repeated_capability'] is True:
+            definition = 'ctypes.create_string_buffer(self._repeated_capability.encode(self._encoding))  # case 2'
+        elif parameter['type'] == 'ViChar':
+            definition = 'ctypes.create_string_buffer({0}.encode(self._encoding))  # case 3'.format(parameter['python_name'])
+        elif parameter['is_buffer'] is True:
+            definition = '(visatype.{0} * len({1}))(*{1})  # case 4'.format(parameter['ctypes_type'], parameter['python_name'], parameter['python_name'])
```
[the remark this answers is not in the document]
Comment (marcoskirsch) [reply build/helper/codegen_helper.py]: I was very confused by your comment until I realized Markdown was rendering the asterisk as italic (I've edited and escaped the asterisks).

[the remark this answers is not in the document]
Comment (marcoskirsch) [reply build/helper/codegen_helper.py]: It is needed in some cases.

```
================================================================================================================== FAILURES ==================================================================================================================
___________________________________________________________________________________________________ TestSession.test_array_input_function ____________________________________________________________________________________________________

self = <test_session.TestSession object at 0x10e3c79d0>

    def test_array_input_function(self):
        test_array = [1, 2, 3, 4]
        test_array_size = len(test_array)
        self.patched_library.niFake_ArrayInputFunction.side_effect = self.side_effects_helper.niFake_ArrayInputFunction
        with nifake.Session('dev1') as session:
>           session.array_input_function(test_array)

bin/nifake/nifake/tests/test_session.py:285: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <nifake.session.Session object at 0x10e446990>, an_array = [1, 2, 3, 4]

    def array_input_function(self, an_array):
        '''array_input_function
    
            This function takes an array parameter.
    
            Args:
                number_of_elements (int): Number of elements in the array.
                an_array (list of float): Contains an array of float numbers
            '''
        vi_ctype = visatype.ViSession(self._vi)  # case 1
        number_of_elements_ctype = visatype.ViInt32(len(an_array))  # case 5
>       an_array_ctype = (visatype.ViReal64 * len(an_array))(an_array)  # case 4
E       TypeError: a float is required

bin/nifake/nifake/session.py:487: TypeError
___________________________________________________________________________________________________ TestSession.test_multiple_array_types ____________________________________________________________________________________________________

self = <test_session.TestSession object at 0x10e5ee310>

    def test_multiple_array_types(self):
        self.patched_library.niFake_MultipleArrayTypes.side_effect = self.side_effects_helper.niFake_MultipleArrayTypes
        passed_in_array = [0.0, 1.0]
        passed_in_array_size = len(passed_in_array)
        fixed_size_array = [2.0, 3.0, 4.0]
        len_array = [5.0, 6.0, 7.0, 8.0]
        self.side_effects_helper['MultipleArrayTypes']['passedInArray'] = passed_in_array
        self.side_effects_helper['MultipleArrayTypes']['aFixedArray'] = fixed_size_array
        self.side_effects_helper['MultipleArrayTypes']['return'] = 0
        with nifake.Session('dev1') as session:
>           passed_in_array_result, fixed_size_array_result = session.multiple_array_types(passed_in_array_size, len_array)

bin/nifake/nifake/tests/test_session.py:343: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <nifake.session.Session object at 0x10e6dced0>, passed_in_array_size = 2, len_array = [5.0, 6.0, 7.0, 8.0]

    def multiple_array_types(self, passed_in_array_size, len_array):
        '''multiple_array_types
    
            Returns multiple types of arrays.
    
            Args:
                passed_in_array_size (int): Number of measurements to acquire.
                len_array_size (int): Size of lenArray
                len_array (list of float): Contains an array of float numbers.
    
            Returns:
                passed_in_array (list of float): An array with size passed in.
    
                    Note: The size must be at least arraySize.
                a_fixed_array (list of float): An array of doubles with fixed size.
            '''
        passed_in_array_size_ctype = visatype.ViInt32(passed_in_array_size)  # case 7
        passed_in_array_ctype = (visatype.ViReal64 * passed_in_array_size)()  # case 12
        a_fixed_array_ctype = (visatype.ViReal64 * 3)()  # case 10
        len_array_size_ctype = visatype.ViInt32(len(len_array))  # case 5
>       len_array_ctype = (visatype.ViReal64 * len(len_array))(len_array)  # case 4
E       TypeError: a float is required

bin/nifake/nifake/session.py:716: TypeError
==================================================================================================== 2 failed, 65 passed in 7.57 seconds =====================================================================================================
```

File: src/nifake/tests/test_session.py (hunk 4 above)
Comment (marcoskirsch) [reply src/nifake/tests/test_session.py]: Moved to matchers.py

File: src/nifake/tests/test_session.py (hunk 1 above)
Comment (marcoskirsch) [reply src/nifake/tests/test_session.py]: Fixed.

File: src/nifake/tests/test_session.py (hunk 3 above)
Comment (marcoskirsch) [reply src/nifake/tests/test_session.py]: Checking that element is of type byte.

File: build/templates/matchers.py.mako [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            return False
+        return True
+
+
+# Scalars
+
+
+class ViBooleanMatcher(_ScalarMatcher):
+    def __init__(self, expected_value):
+        _ScalarMatcher.__init__(self, visatype.ViBoolean, 1 if expected_value is True else False)
```
Comment (texasaggie97) [inline build/templates/matchers.py.mako]: You compare against 1 if expected_value is True, but compare against False when expected_value is False.

File: build/templates/matchers.py.mako [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+# Base classes
+
+
+class _ScalarMatcher(object):
+    def __init__(self, expected_type, expected_value):
+        self.expected_type = expected_type
+        self.expected_value = expected_value
+
+    def __repr__(self):
```
Comment (texasaggie97) [inline build/templates/matchers.py.mako]: Should all of the matchers have a `__repr__` or is the default good enough for the rest?

File: build/templates/matchers.py.mako [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+# Buffers
+
+
+class ViBooleanBufferMatcher(_BufferMatcher):
+    def __init__(self, expected_size_or_value):
+        _BufferMatcher.__init__(self, visatype.ViBoolean, expected_size_or_value)
+
+
+class ViSessionBufferMatcher(_BufferMatcher):
```
Comment (texasaggie97) [inline build/templates/matchers.py.mako]: Do we need a ViSession buffer matcher?

File: build/templates/matchers.py.mako (hunk 8 above)
Comment (marcoskirsch) [reply build/templates/matchers.py.mako]: Now that I see it, that __repr__ implementation is actually outdated. I lean towards removing it, errors were easy enough to follow.

File: build/templates/matchers.py.mako (hunk 7 above)
Comment (marcoskirsch) [reply build/templates/matchers.py.mako]: Fixed!

File: build/templates/matchers.py.mako (hunk 9 above)
Comment (marcoskirsch) [reply build/templates/matchers.py.mako]: No need for it yet. I'd rather not create it now.

**Pull request merged**